app/herdbook.py

```diff
@@ -505,6 +509,314 @@ def testbreed():
+    """
+    return s3.get_s3_client().head_object(object_name=f"{ind_number}/certificate.pdf")
+
+
+def verify_certificate_checksum(ind_number, checksum):
+    """
+    Returns whether a certificate exists with the given checksum.
+    """
+    s3_sum = get_certificate_checksum(ind_number)
+    return s3_sum == checksum
+
+
+def flatten_list_of_dcts(in_list):
+    """
+    Flattens a list of dictionaries
+    """
+    dct = {}
+    for item in in_list:
+        if item is not None:
+            dct.update(item)
+    return dct
+
+
+def get_certificate_data(ind, user_id):
+    """
+    Gets all data needed to issue a certificate.
+    """
+    parent_keys = [
+        (1, "F"),
+        (1, "M"),
+        (2, "FF"),
+        (2, "MF"),
+        (2, "FM"),
+        (2, "MM"),
+    ]
+
+    date = datetime.datetime.utcnow()
+    date = date.strftime("%Y-%m-%d")
+    extra_data = {"user_id": user_id, "issue_date": date, "photos": False}
+    ind["notes"] = "\n".join(
+        [
+            "Notes: " + str(ind.get("notes", "")),
+            "Color notes: " + str(ind.get("color_note", "")),
+            "Hair notes: " + str(ind.get("hair_notes", "")),
+        ]
+    )
+    cert_data_lst = []
+    cert_data_lst.append(ind)
+    cert_data_lst.append(extra_data)
+    for level, a_type in parent_keys:
+        cert_data_lst.append(_get_parent(ind, user_id, level, a_type))
+
+    return flatten_list_of_dcts(cert_data_lst)
+
+
+def _get_parent(ind, user_id, ancestry_level, ancestry_type):
+    ancestries = {
+        (1, "F"): "father",
+        (1, "M"): "mother",
+        (2, "FF"): "father",
+        (2, "MF"): "father",
+        (2, "FM"): "mother",
+        (2, "MM"): "mother",
+    }
+    try:
+        if ancestry_level == 1:
+            ancestor = ind.get(ancestries[(1, ancestry_type)], None)
+            idv = da.get_individual(ancestor["number"], user_id)
+
+        elif ancestry_level == 2:
+            ancestor = ind.get(ancestries[(1, ancestry_type[0])], None)
+            parent = da.get_individual(ancestor["number"], user_id)
+            grand_ancestor = parent.get(ancestries[(2, ancestry_type)], None)
+            idv = da.get_individual(grand_ancestor["number"], user_id)
+
+    except TypeError as ex:
+        print(ex)
+        return None
+
+    ndict = dict()
+    for key in idv.keys():
+        ndict[key] = ancestry_type + "_" + key
+
+    parent_ind = dict()
+    for old_key, val in idv.items():
+        parent_ind[ndict[old_key]] = val
+
+    return parent_ind
+
+
+def get_certificate(data):
+    """
+    Returns a pdf certificate of an individual.
+    """
+    # pylint: disable=R0914
+    qr_x_pos, qr_y_pos = 295, 795
+    # version 2 means size 42x42
+    qr_x_len, qr_y_len = 42, 42
+
+    certificate = certs.CertificateGenerator(
+        form=settings.certs.template,
+        form_keys=certs.FORM_KEYS,
+    )
+
+    qr_code = certs.QRHandler(
+        link=settings.service.host + "/individual/" + data["number"] + "/verify",
```

Comment:
I'm fine with merging this for now, but think the host should really be picked up from the request (likely requires adding 
`proxy_set_header Host $host;` in nginx).